core/domain/classifier_services.py

```diff
@@ -347,6 +348,75 @@ def mark_training_job_complete(job_id):
                                            feconf.TRAINING_JOB_STATUS_COMPLETE)
 
 
+def mark_training_job_failed(job_id):
+    """Updates the training job's status to failed.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob.
+    """
+    _update_classifier_training_job_status(job_id,
+                                           feconf.TRAINING_JOB_STATUS_FAILED)
+
+
+def mark_training_job_pending(job_id):
+    """Updates the training job's status to pending.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob.
+    """
+    _update_classifier_training_job_status(job_id,
+                                           feconf.TRAINING_JOB_STATUS_PENDING)
+
+
+def update_failed_jobs(training_jobs):
+    """Updates status of job_models to failed.
+
+    Args:
+        training_jobs: list. List of jobs to be updated as failed.
+    """
+    for training_job in training_jobs:
+        mark_training_job_failed(training_job.job_id)
+
+
+def fetch_next_job():
+    """Gets next job model in the job queue.
+
+    Returns:
+        ClassifierTrainingJobModel. Domain object of the next training Job.
+    """
+    classifier_training_jobs = []
+    cursor = None
+    valid_jobs = []
+    failed_jobs = []
+
+    while len(valid_jobs) == 0:
+        classifier_training_jobs, cursor, more = (
+            classifier_models.ClassifierTrainingJobModel.query_training_jobs(
+                cursor))
+        for training_job in classifier_training_jobs:
+            if(training_job.status == (
+                    feconf.TRAINING_JOB_STATUS_PENDING)):
+                if (datetime.datetime.utcnow() - (
+                        training_job.last_updated) > (
+                            datetime.timedelta(
+                                minutes=feconf.CLASSIFIER_JOB_TTL))):
+                    failed_jobs.append(training_job)
+                else:
+                    valid_jobs.append(training_job)
```

Comment:
I don't think this qualifies as valid job. If job is pending and timeout hasn't happened yet then it is pending job. It shouldn't be considered as next job for training. Am I missing something here?